README.md
[![Maintainability](https://api.codeclimate.com/v1/badges/006b4b12feb004d15164/maintainability)](https://codeclimate.com/github/iTonyYo/watermark-cli/maintainability)

# `@oopsunome/watermark-cli`

从命令行给图片贴水印。

## 目录
- [安装](#安装)
- [使用](#使用)
- [相关](#相关)
- [贡献指南](#贡献指南)
- [证书](#证书)

## 安装

```shell
# 使用 NPM
$ npm i -g @oopsunome/watermark-cli

# 使用 Yarn
$ yarn global add @oopsunome/watermark-cli
```

## 使用

```
$ watermark <标记> <画布> 选项 [...]

  选项
    --out, -o, 输出位置, 默认 process.cwd()
    --scale, -s, 水印缩放程度, 0 ~ 1
    --opacity, -O, 水印透明度, 0 ~ 1
    --position, -p, 水印位置, center | top-left | top-right | bottom-left | left-bottom
    --gap, -g, 画布内边距, 默认 8 像素

  示例
    $ watermark mark.png paper.png --scale 0.3 --opacity 0.8 --position center --gap 18
```

## 相关
- [@oopsunome/watermark][@oopsunome/watermark] - 当前项目调用的接口。

## 贡献指南

仔细查阅 [CONTRIBUTING.md][贡献指南] 以了解详情。

## 证书

[`@oopsunome/watermark-cli`][@oopsunome/watermark-cli] 获得了 MIT 许可，仔细查阅 [LICENSE.md][证书] 以了解详情。



[贡献指南]: https://github.com/iTonyYo/watermark-cli/blob/master/CONTRIBUTING.md
[证书]: https://github.com/iTonyYo/watermark-cli/blob/master/LICENSE.md
[Node]: https://nodejs.org/
[@oopsunome/watermark-cli]: https://github.com/iTonyYo/watermark-cli
[@oopsunome/watermark]: https://github.com/iTonyYo/watermark


CHANGELOG.md
# [`@oopsunome/watermark-cli`](https://github.com/iTonyYo/watermark-cli) 更新日志

## [v1.0.5](https://github.com/iTonyYo/watermark-cli/compare/v1.0.4...v1.0.5) - 2019-08-30

#### 提交

- [[`58e3055`](https://github.com/iTonyYo/watermark-cli/commit/58e3055e0352b7de3ca06a3c5de0f2b311737996)] 依赖: :arrow_up: 升级 `@oopsunome/watermark`
## [v1.0.4](https://github.com/iTonyYo/watermark-cli/compare/v1.0.3...v1.0.4) - 2019-08-28

#### 提交

- [[`d157fee`](https://github.com/iTonyYo/watermark-cli/commit/d157feefefa6c562f965e0ccf1cae734f4b56f98)] 依赖: :arrow_up: 更新依赖
- [[`99430db`](https://github.com/iTonyYo/watermark-cli/commit/99430dbd325edfaf4b0c10da3d484cb26b1fb543)] 配置: :wrench: 更新若干配置
- [[`aab7979`](https://github.com/iTonyYo/watermark-cli/commit/aab79793ad9ed8c84c1423fc8a61a0d6333daa6d)] 工作流: :railway_track: 更新自动化指令
- [[`3f3528d`](https://github.com/iTonyYo/watermark-cli/commit/3f3528dece5c20453905467f52f66093af231e32)] 文档: :memo: 更新 README.md
- [[`14b95c0`](https://github.com/iTonyYo/watermark-cli/commit/14b95c01687e2cf1841db688d83cf0de6f67eab1)] 发版: :bookmark: v1.0.4
- [[`1b677ca`](https://github.com/iTonyYo/watermark-cli/commit/1b677ca08e7f8a48aba40e22b22b350f8dd26748)] 文档: :memo: 新增拉去请求 & 问题模板
- [[`dc83e66`](https://github.com/iTonyYo/watermark-cli/commit/dc83e66d3ac75eba8147553dc3734f8ee805a5b5)] 功能: :gear: 通过 `-h` 化名使用 `--help` 选项
## [v1.0.3](https://github.com/iTonyYo/watermark-cli/compare/v1.0.2...v1.0.3) - 2019-01-09

#### 提交

- [[`44d5275`](https://github.com/iTonyYo/watermark-cli/commit/44d527524dc34c9e542ba131e002b31f4c1fa705)] 工程: :clapper: 配置更新日志
- [[`28d7b99`](https://github.com/iTonyYo/watermark-cli/commit/28d7b99f8fc99b31a937e3cd56d233e0a5cede75)] 文档: :memo: 更新文档
- [[`c6285b4`](https://github.com/iTonyYo/watermark-cli/commit/c6285b426a162ccd423e66c6546e38a3a970f55b)] 工程: :clapper: 升级依赖
## [v1.0.2](https://github.com/iTonyYo/watermark-cli/compare/v1.0.1...v1.0.2) - 2019-01-07

#### 提交

- [[`a5ddf15`](https://github.com/iTonyYo/watermark-cli/commit/a5ddf15473f3336f08ae1f22fcace235cd8c7758)] 工程: :clapper: 升级依赖
## [v1.0.1](https://github.com/iTonyYo/watermark-cli/compare/v1.0.0...v1.0.1) - 2019-01-07

#### 提交

- [[`d83f9d7`](https://github.com/iTonyYo/watermark-cli/commit/d83f9d7f4c6b8615fca8437ee0a174a6fcd1c97f)] 文档: :memo: 更新文档
## v1.0.0 - 2019-01-07

#### 提交

- [[`4c52ecc`](https://github.com/iTonyYo/watermark-cli/commit/4c52ecc5ddad286ceff414a71c256c62a34ac7d4)] 新功能: :tada: 所有选项支持化名
- [[`a026542`](https://github.com/iTonyYo/watermark-cli/commit/a026542013985a21640f5579489869992b07b65b)] 工程: :clapper: 重新生成 `licenses-*.csv`
- [[`2660ea0`](https://github.com/iTonyYo/watermark-cli/commit/2660ea05ba50f274b7c1cb5c7554edd46f5bb10b)] 文档: :memo: 更新 `使用` 指南


scripts/license.js
#!/usr/bin/env node
'use strict';

const { join }        = require('path');
const licenseChecker  = require('license-checker');
const makeDir         = require('make-dir');
const writeFileAtomic = require('write-file-atomic');
const chalk           = require('chalk');

const { appDirectory, docsPath } = require('./util');

/**
 * 异步保存报告，
 *  - 保存的位置是否存在
 *  - 保存文件
 *
 * @params {Object} csv - CSV 数据
 * @params {String} to - 保存的位置
 * @params {String} fileName - 报告文件名称
 */
const saveCSV  = ({csv, to, fileName}) => {
  makeDir(to)
    .then(path => {
        writeFileAtomic(
          join(to, fileName),
          `${licenseChecker.asCSV(csv)}\n`,
          {},
          (err) => {
            if (!err) {
              console.log(`
                ${chalk.green.bold(`${fileName}`)}
                ${chalk.dim('保存在: ') + chalk.blueBright(path)}
              `);

              return;
            }

            throw err;
          }
        );
      })
    .catch(err => {
      throw err;
    });
};

/**
 * 生成 package.json 文件中 "dependencies" 依赖证书报告
 */
licenseChecker.init({
  start:       appDirectory,
  production:  false,
  development: true,
}, (err, data) => {
  if (err) {
    throw err;
  } else {
    saveCSV({
      csv     : data,
      to      : docsPath,
      fileName: 'licenses-development.csv',
    });
  }
});

/**
 * 生成 package.json 文件中 "devDependencies" 依赖证书报告
 */
licenseChecker.init({
  start:       appDirectory,
  production:  true,
  development: false,
}, (err, data) => {
  if (err) {
    throw err;
  } else {
    saveCSV({
      csv     : data,
      to      : docsPath,
      fileName: 'licenses-production.csv',
    });
  }
});


cli.js
#!/usr/bin/env node

const {
  dirname,
  basename,
} = require('path');

const meow = require('meow');
const updateNotifier = require('update-notifier');
const { isString, merge } = require('lodash');

const watermark = require('@oopsunome/watermark');

const cli = meow(`
  使用方式
    $ watermark <mark> <paper> options [...]

  选项
    --out, -o, 输出位置, 默认 process.cwd()
    --scale, -s, 水印缩放程度, 0 ~ 1
    --opacity, -O, 水印透明度, 0 ~ 1
    --position, -p, 水印位置, center | top-left | top-right | bottom-left | left-bottom
    --gap, -g, 画布内边距, 默认 8 像素

  示例
    $ watermark mark.png paper.png --scale 0.3 --opacity 0.8 --position center --gap 18
`, {
  flags: {
    out: {
      type: 'string',
      alias: 'o',
    },
    scale: {
      type: 'number',
      alias: 's',
    },
    opacity: {
      type: 'number',
      alias: 'O',
    },
    position: {
      type: 'string',
      alias: 'p',
    },
    gap: {
      type: 'number',
      alias: 'g',
    },
    help: {
      type: 'boolean',
      alias: 'h',
    },
  },
});

updateNotifier({ pkg: cli.pkg }).notify();

if (cli.input.length !== 2) {
  throw Error('必须同时提供水印、画板两个来源');
}

(async () => {
  const { input, flags } = cli;
  const {
    gap,

    position,
    scale,
    opacity,

    out,
  } = flags;

  await watermark(
    input[0],
    input[1],
    merge(
      {
        gap,
        mark: {
          position,
          scale,
          opacity,
        },
        output: {
          filename: undefined,
          path: undefined,
        },
      },
      isString(out) ? {
        output: {
          filename: basename(out),
          path: dirname(out),
        },
      } : {
        output: {
          filename: undefined,
          path: undefined,
        },
      },
    ),
  );
})();
